### DISCUSSION TITLE
Are there any examples of unit tests with Multer ?

### DISCUSSION THREAD

**Discussion by AlexMachin1997** [expressjs/multer#1196] [Q&A]
Hi everyone,

I've just finished adding Multer to my project, however, now I've got write some unit tests.

Is there any easy way to mock how Multer work's without adding a bunch of complexity ?

**I want to be able to test the following:**
- File type validation
- Successful upload
- Unexpected system errors

**Comment by mrdulin** [comment]
Can you provide a minimal code you want to test?

**Comment by Tommy1223454** [comment]
If it helps, I’d avoid mocking Multer too much and instead test the upload route with real fixture files.

The useful cases are usually:
- a valid expected file
- a renamed file with a misleading extension
- MIME/extension mismatch cases
- a small known-bad fixture like EICAR
- archive edge cases if your pipeline accepts ZIPs

I built an OSS package around this inspect-before-storage step:
https://github.com/pompelmi/pompelmi

Even if you don’t use it directly, that kind of fixture-driven testing tends to be much more valuable than only mocking the happy path.

**Comment by Errr0rr404** [comment]
You usually do **not** need to mock Multer’s internals. Drive it through HTTP with `supertest` (or call the middleware with a crafted `req`).

### 1) Integration-style (recommended)

```js
const request = require("supertest");
const express = require("express");
const multer = require("multer");
const upload = multer({
  storage: multer.memoryStorage(),
  fileFilter(req, file, cb) {
    if (file.mimetype !== "image/png") return cb(null, false);
    cb(null, true);
  },
});

const app = express();
app.post("/upload", upload.single("file"), (req, res) => {
  if (!req.file) return res.status(400).json({ error: "bad type" });
  res.json({ ok: true, size: req.file.size });
});

// success
await request(app)
  .post("/upload")
  .attach("file", Buffer.from("fake"), { filename: "a.png", contentType: "image/png" })
  .expect(200);

// rejected type
await request(app)
  .post("/upload")
  .attach("file", Buffer.from("fake"), { filename: "a.txt", contentType: "text/plain" })
  .expect(400);
```

### 2) Error path
Multer errors are often passed to `next(err)` with codes like `LIMIT_FILE_SIZE`. Add an Express error middleware and assert on `err.code`.

### 3) Avoid real disk
Use `multer.memoryStorage()` in tests so nothing hits the filesystem.

That covers type validation, happy path, and system/limit errors without deep mocks.
